address.md

```diff
@@ -0,0 +1,403 @@
+* Most of the research on error-detecting codes only applies to character-set sizes that are a [prime power](https://en.wikipedia.org/wiki/Prime_power), which 58 is not.
+* Base58 decoding is complicated and relatively slow.
+
+Learning from the Ethereum:
+
+* Hex encoding is inefficient with respect to storage. Representing addresses as Hex increases the address size by 100%.
+* Checksums require mixed cases, resulting in the same issues as above.
+
+## Specification
+
+A Filecoin address is a byte string consisting of:
+* The **letter 'f'** to identify the address as belonging to a Filecoin network.
+* A **network indicator** that identifies which network this address belongs to.
+* A **protocol indicator** that identifies the type and version of this address.
+* The **data** used to uniquely identify the actor according to the protocol.
+
+```
+|---|---------|----------|------|
+| f | network | protocol | data |
+|---|---------|----------|------|
+```
+
+An example of a Filecoin address in golang,
+
+```go
+type Address struct {
+    // 0: main-net
+    // 1: test-net
+    network varint
+    
+    // 0: Blake2b160-Hash of SECP256K1 Public Key
+    // 1: ActorID
+    // 2: BLS Public Key -> in bytes?
+    protocol varint
+    
+    // raw bytes containing the data associated with typ
+    data []byte
+}
+```
+
+#### Network Indicator
+
+The **network** is a `var32encoded` integer that identifies the network:
+
+- `0` :  Mainnet
+- `1` : Testnet 1
+- `2` : Testnet 2
+- `3` : etc.
+
+The size of this integer is `4` bytes. An example description in golang,
+
+> Note(bvo): Double check that this is the integer 4 bytes?
+
+```go
+// String Prefix
+const (
+    MAINNET_PREFIX = "fc"
+    TESTNET_PREFIX = "tf"
+)
+
+// Network Byte
+const (
+	Mainnet = 0
+	Testnet = 1
+)
+```
+
+#### Protocol Indicator
+
+The **protocol indicator** is a *var32encoded* integer that describes how a method should interpret the information in the `data` field of an address. This indicator  specifies the hashing algorithms, error correcting protocols, and the origin of the hashed data (e.g. a public key) when needed. Any deviation for the algorithms and data types specified by the protocol must be assigned a new protocol number. In this way, protocols also act as versions.
+
+```go
+// Type varint
+const (
+	SECP256K1 = 0
+	ID = 1
+	Actor = 2
+	BLS = 3
+)
+```
+
+###### Protocol 0: libsecpk1 Elliptic Curve Public Keys
+
+**Protocol 0** addresses represent public encryption keys. These addresses represent Filecoin user accounts and are likely to be transmitted outside the network over noisy channels (e.g. hand written on a napkin). As such, they require robust error detection. The multi-base encoded key contains a hash and a checksum. The hash is a [Blake2b 160](https://blake2.net/) hash of the public key. The checksum is a six character BCH checksum. The algorithm details can be found in Appendix - 9.2. Algorithms - Checksums.
```

Comment:
I believe we either need to add a second protocol for the case of Actors without public keys, or we need to state that the address represent public encryption keys **or** "random data" used to generate an address. I think "random data" is a pretty poor description, thoughts?

Reply:
Also, I believe we decided that we are **not** using multi-base encoding, correct?